lib/recommendations.js

```diff
@@ -0,0 +1,171 @@
+  } catch (err) {
+    // Ignore NotFoundError, this means we need to create instead of update
+    if (!(err instanceof httph.NotFoundError)) {
+      throw err;
+    }
+  }
+
+  if (!recommendation) {
+    return insert_recommendation(pg_pool, [uuid.v4(), app_uuid, service, resource_type, action, details]);
+  }
+
+  return update_recommendation(pg_pool, [recommendation.recommendation_uuid, details]);
+}
+
+/**
+ * HTTP handler for recommendation creation. This should handle creating OR updating
+ * @param {*} pg_pool - Database connection pool
+ * @param {*} req - Request object
+ * @param {*} res - Response object
+ * @param {*} regex - URL path regex
+ */
+async function http_create(pg_pool, req, res, regex) {
+  const payload = await httph.buffer_json(req);
+  const app_key = httph.first_match(req.url, regex);
+
+  // Make sure app exists
+  const app = await common.app_exists(pg_pool, app_key);
+
+  try {
+    assert.ok(payload.resource_type, 'Missing resource_type property - A resource type must be specified when creating a new recommendation');
+    const type = await check_resource_type_exists(pg_pool, payload.resource_type);
+    payload.resource_type = type;
+    await check_payload(pg_pool, payload);
+  } catch (err) {
+    throw new common.UnprocessibleEntityError(err.message);
+  }
+
+  const result = (await create_or_update(
+    pg_pool,
+    app.app_uuid,
+    payload.service,
+    payload.resource_type.resource_type_uuid,
+    payload.action,
+    payload.details,
+  ))[0];
+
+  return httph.created_response(
+    res,
+    JSON.stringify({
+      id: result.recommendation,
+      app: {
+        name: app.app_name,
+        key: `${app.app_name}-${app.space_name}`,
+        id: app.app_uuid,
+      },
+      service: result.service,
+      resource_type: payload.resource_type.name,
+      action: result.action,
+      details: result.details,
+    }),
+  );
+}
+
+/**
+ * HTTP handler for recommendation fetching (stub)
+ * @param {*} pg_pool - Database connection pool
+ * @param {*} req - Request object
+ * @param {*} res - Response object
+ * @param {*} regex - URL path regex
+ */
+// async function http_get(pg_pool, req, res, regex) {}
+
+/**
+ * HTTP handler for recommendation listing (stub)
+ * @param {*} pg_pool - Database connection pool
+ * @param {*} req - Request object
+ * @param {*} res - Response object
+ * @param {*} regex - URL path regex
+ */
+// async function http_list(pg_pool, req, res, regex) {}
+
+/**
+ * HTTP handler for getting available recommendation resource types
+ * @param {*} pg_pool - Database connection pool
+ * @param {*} req - Request object
+ * @param {*} res - Response object
+ */
+async function http_get_resource_types(pg_pool, req, res) {
+  return httph.ok_response(res, JSON.stringify(await get_resource_types(pg_pool)));
+}
+
+module.exports = {
+  // get,
+  // list,
+  create: create_or_update,
```

Comment:
It doesn't look like the `create_or_update` function is used outside of this module. Is there a reason to export it?

Reply:
Good point - I think I had that in there just in case some other module wanted to use it, but we could just export it in the future if that is the case. I'll un-export it